nixos/modules/services/networking/openconnect.nix

```diff
@@ -0,0 +1,154 @@
+        description = ''
+          File containing the password to authenticate with. This
+          is passed to <code>openconnect</code> via the
+          <code>--passwd-on-stdin</code> option.
+        '';
+        default = null;
+        example = "/var/lib/secrets/openconnect-passwd";
+        type = types.path;
+      };
+
+      certificate = mkOption {
+        description = ''
+          Certificate to authenticate with. Either a file name or
+          a PKCS#11 URL.
+        '';
+        default = null;
+        example = "/var/lib/secrets/openconnect_certificate.pem";
+        type = types.nullOr types.str;
+      };
+
+      privateKey = mkOption {
+        description = ''
+          Private key to authenticate with. Either a file name or
+          a PKCS#11 URL.
+        '';
+        example = "/var/lib/secrets/openconnect_private_key.pem";
+        default = null;
+        type = types.nullOr types.str;
+      };
+
+      extraConfig = mkOption {
+        description = ''
+          Extra config to be appended to the interface config. It should
+          contain long-format options as would be accepted on the command
+          line by <code>openconnect</code> (see https://www.infradead.org/openconnect/manual.html),
+          separated with newlines. Empty lines and lines where the first
+          non-space character is # are ignored.
+        '';
+        default = "";
+        example = ''
+          # Set compression mode
+          compression=stateless
+
+          no-http-keepalive
+          no-dtls
+        '';
+        type = types.str;
+      };
+    };
+  };
+  generateConfig = name: icfg: pkgs.writeText "config" ''
+    interface=${name}
+    ${optionalString (icfg.user != null) "user=${icfg.user}"}
+    ${optionalString ((icfg.password != null) || (icfg.passwordFile != null)) "passwd-on-stdin"}
+    ${optionalString (icfg.certificate != null) "certificate=${icfg.certificate}"}
+    ${optionalString (icfg.privateKey != null) "sslkey=${icfg.privateKey}"}
+
+    ${icfg.extraConfig}
+  '';
+  generateUnit = name: icfg:
+    let
+      # password can only be passed via stdin, so we also need a password file if the password option is used
+      needsPasswordFile = (icfg.password != null) || (icfg.passwordFile != null);
+      passwordFile = if (icfg.password != null)
+        then (pkgs.writeText "password" icfg.password)
+        else icfg.passwordFile;
+    in {
+      description = "OpenConnect Interface -  ${name}";
+      requires = [ "network-online.target" ];
+      after = [ "network.target" "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        Type = "simple";
+        ExecStart = if needsPasswordFile
+          then "${pkgs.bash}/bin/sh -c 'cat ${passwordFile} | ${openconnect}/bin/openconnect --config=${generateConfig name icfg} ${icfg.gateway}'"
+          else "${openconnect}/bin/openconnect --config='${generateConfig name icfg} ${icfg.gateway}'";
```

Comment:
systemd has a `StandardInput` directive than can be pointed to file, using that is probably better than going through a shell. (not entirely sure how this interacts with `LoadCredential`, that might be worth looking too)

Reply:
I think we don't need `LoadCredential` here. As far as I understood, `LoadCredential` provides secrets to the process via the filesystem, however here it should be sufficient to just let `StandardInput` read directly from the specified location as we need to use it anyway

Reply:
we don't need it, but it would've been nice to use if it's easily possible. that way hardening the service later on (which we haven't looked at yet) might become easier. but for now it is totally fine without :)

Reply:
That's definitely a good idea - this module should be fairly easy to harden because it is stateless and just needs network access